### DISCUSSION TITLE
Best measure approach

### DISCUSSION THREAD

**Discussion by jlvila** [jimy-byerley/pymadcad#28] [Q&A] [answered]
Hi,

Context: PyMadCad 0.11, Python 3.9 and W10

I just want to measure a distance between 2 points that come from a "mouseReleaseEvent".
First of all, I need to convert 2D coordinates into 3D real coordinates and I use ptat() / or ptfrom(). And to materialized the distance between these points, I simply use a note_distance() **=> 206**.
None of the approaches works as expected: ptat(), ptfrom(..., O), ptfrom(.., _mesh.box().min)
The first view corresponding to the scene just after clicks, and the second when I navigate elsewhere and as you can see, the note_distance is not well positionned.
![measure1](https://user-images.githubusercontent.com/8035931/157902746-a5e8c8ea-b5e8-4b74-925a-97903660c6f8.png)
![measure_change_point_of_view](https://user-images.githubusercontent.com/8035931/157902759-0ad74029-be5b-4929-b5aa-37a8d40e8c68.png)

**Comment by jimy-byerley** [comment]
Hello again

A combination of `View.ptat` and `View.somenear` should be the solution for this. 
I think what you are looking for is exactly [this](https://github.com/jimy-byerley/pymadcad/blob/07adf336a67faac606b4575774a4b011d0cc658d/madcad/kinematic.py#L239) isn't it ?
And this should give something quite close to the pointed position on the surface. however this sadly won't be precisely *on* the surface because of a known bug I still haven't fixed (there can be error of ~5% of the view-object distance).  

If it does not work for you, can you submit a piece of code ?

### For more details about the precision issue

I rely on the OpenGL internal depthmap to get the view-object distance, and then recover the 3d position with the view perspective informations. But for some reasons OpenGL is giving me incorrect values (not conform to what the specifications I read says). So I [experimentally patched the wrong depth value](https://github.com/jimy-byerley/pymadcad/blob/07adf336a67faac606b4575774a4b011d0cc658d/madcad/rendering.py#L805) but the result is not precise as it should have been.

**Comment by jlvila** [comment]
Jimy,

I think I miss something but I don't know what. I understand the precision issue but, here is my example.

- I click 2 points into my View (upper left corner and upper right), and they seem well positionned
![FirstScreen](https://user-images.githubusercontent.com/8035931/158365757-847480ac-c509-4b8d-aa1c-950327a94a43.png)
- Then I navigate and you can see that the points are not well positionned!
![secondPointOfView](https://user-images.githubusercontent.com/8035931/158366045-e9095705-6cc0-4b74-8fc2-09c80250b2f1.png)

here my piece of code

```python
        if isinstance(evt, QMouseEvent) and evt.type() in (QEvent.MouseButtonRelease, QEvent.MouseButtonDblClick):
            pos = self.somenear(evt.pos(), radius=2)
            if pos:
                key = self.itemat(pos)
                _k = key[0]
                if type(key[0]) == type(""):
                    _k = int(key[0].split('_', 1)[0])
                if not _k in self.transport.shapes3d:
                    return
                self.transport.scene3d.refreshmaps()
                _mesh = self.transport.shapes3d[_k]
                if self.firstPoint is None:
                    self.firstPoint = self.ptat(pos)
                    print(
                        f'DISTANCE - Point 1: {self.ptfrom(pos, _localOrigin)}    /// ptat {self.ptat(pos)}')
                else:
                    print(
                        f'DISTANCE - Point 2: {self.ptfrom(pos, _localOrigin)}    /// ptat {self.ptat(pos)}')
                    _secondPoint = self.ptat(pos)
                    firstPointG = Displayable(
                        PointDisplay, position=self.firstPoint, color=(1.0, 0, 0, 1.0), )
                    secondPointG = Displayable(
                        PointDisplay, position=_secondPoint, color=(1.0, 0, 0, 1.0), )
                    nd = note_distance(
                        self.firstPoint, _secondPoint, text='d: {d: .4g} ±{tol}', tol=0.05)
                    self.firstPoint = None

```

**Reply by jimy-byerley** [reply]
I played with `ptat` a bit yesterday, and I found that changing the `0.53` factor into `0.5` now gives a very good result, which was not the case when I wrote that function. So the problem is even more odd than I expected !
I have to investigate on this.

If you really need that function working for now, you can change the factor in the code of your pymadcad installation (which is ugly by I don't have a better solution at the moment).

**Reply by jimy-byerley** [reply]
I opened an issue about this, you will be able to see advancement at #32 
Thanks for reporting this problem !

**Comment by jlvila** [comment, marked as answer]
Jimy, 
I just change 0.53 to 0.5 and the result is really better :-)
thanks again
Jean-Louis
